examples/research_projects/codeparrot/scripts/preprocessing.py

```diff
@@ -79,45 +82,58 @@ def filter(example, uniques, args):
-args = parser.parse_args()
-if args.num_workers is None:
-    args.num_workers = multiprocessing.cpu_count()
-
-# Load dataset
-t_start = time.time()
-ds = load_dataset(args.dataset_name, split="train")
-print(f"Time to load dataset: {time.time()-t_start:.2f}")
-
-# Run preprocessing
-t_start = time.time()
-ds = ds.map(preprocess, num_proc=args.num_workers)
-print(f"Time to preprocess dataset: {time.time()-t_start:.2f}")
-
-# Deduplicate hashes
-uniques = set(ds.unique("hash"))
-frac = len(uniques) / len(ds)
-print(f"Fraction of duplicates: {1-frac:.2%}")
-
-# Deduplicate data and apply heuristics
-t_start = time.time()
-ds_filter = ds.filter(filter, fn_kwargs={"uniques": uniques, "args": args})
-print(f"Time to filter dataset: {time.time()-t_start:.2f}")
-print(f"Size of filtered dataset: {len(ds_filter)}")
-
-# Save data in batches of samples_per_file
-if not os.path.exists(args.output_dir):
-    os.makedirs(args.output_dir)
-t_start = time.time()
-for file_number, index in enumerate(range(0, len(ds_filter), args.samples_per_file)):
-    file_path = f"{args.output_dir}/file-{file_number+1:012}.json"
-    end_index = min(len(ds_filter), index + args.samples_per_file)
-    ds_filter.select(list(range(index, end_index))).to_json(file_path)
-    compress_file(file_path)
-print(f"Time to save dataset: {time.time()-t_start:.2f}")
+if __name__ == "__main__":
+    # Settings
+    parser = HfArgumentParser(PreprocessingArguments)
+    args = parser.parse_args()
+    if args.num_workers is None:
+        args.num_workers = multiprocessing.cpu_count()
+
+    # Load dataset
+    t_start = time.time()
+    ds = load_dataset(args.dataset_name, split="train")
+    print(f"Time to load dataset: {time.time()-t_start:.2f}")
+
+    # Run preprocessing
+    t_start = time.time()
+    ds = ds.map(preprocess, num_proc=args.num_workers)
+    print(f"Time to preprocess dataset: {time.time()-t_start:.2f}")
+
+    # Deduplicate hashes
+    uniques = set(ds.unique("hash"))
+    frac = len(uniques) / len(ds)
+    print(f"Fraction of duplicates: {1-frac:.2%}")
+
+    # Deduplicate data and apply heuristics
+    t_start = time.time()
+    ds_filter = ds.filter(filter, fn_kwargs={"uniques": uniques, "args": args})
+    print(f"Time to filter dataset: {time.time()-t_start:.2f}")
+    print(f"Size of filtered dataset: {len(ds_filter)}")
+
+    # Deduplicate with minhash and jaccard similarity
+    t_start = time.time()
+    ds_dedup, duplicate_clusters = deduplicate_dataset(ds_filter)
+    print(f"Time to deduplicate dataset: {time.time()-t_start:.2f}")
+    print(f"Size of deduplicate dataset: {len(ds_dedup)}")
+
+    # Save data in batches of samples_per_file
+    output_dir = Path(args.output_dir)
+    output_dir.mkdir(exist_ok=True)
+
+    # save duplicate_clusters in the output_dir as artifacts
+    # not sure it is the right place the save it
+    with open(output_dir / "duplicate_clusters.json", "w") as f:
+        json.dump(duplicate_clusters, f)
```

Comment:
We will push the `output_dir` to the hub and load it as a dataset later. So we don't want the that to be in the same folder as the data files. A solution is to store the data files in `output_dir / data / f"file-{file_number+1:012}.json"` and keep the duplicate clusters where they are. I think then `datasets` will only load the files in `data/`.

Reply:
done